### ISSUE TITLE
Incompatible with NestJS cache store

### ISSUE THREAD

**Issue by consult-kk** [dabroek/node-cache-manager-redis-store#53]
V3.0.1 is incompatible with NestJS . Unable to register CacheModule using the RedisStore.

	"message": "Type 'typeof import(\"cache-test/node_modules/cache-manager-redis-store/dist/index\")' is not assignable to type '(string | CacheStoreFactory | CacheStore) & typeof import(\"/Users/kk/dev/nodejs/cache-test/node_modules/cache-manager-redis-store/dist/index\")'.\n  ...

Here is the code I am using in the module.ts : 
`
import * as redisStore from "cache-manager-redis-store";
import { ConfigModule } from '@nestjs/config';

@Module({
  imports: [
    ConfigModule.forRoot(),
    CacheModule.register({
      isGlobal: true,
      store: redisStore,
      url: "redis://localhost:6379",
    }),
    HttpModule,
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
`

**Comment by dimaqw** [comment]
The same issue

**Comment by Alfagun74** [comment]
Same Issue

**Comment by belyaev-dev** [comment]
Same issue

**Comment by dougrocha** [comment]
I found a way to fix this issue although it may not be nice. 
```
import { redisStore } from 'cache-manager-redis-store'

CacheModule.registerAsync({
      isGlobal: true,
      imports: [ConfigModule],
      useFactory: async (configService: ConfigService) => ({
        ttl: configService.get('CACHE_TTL'),
        store: (await redisStore({
          url: configService.get('REDIS_URL'),
        })) as unknown as CacheStore,
      }),
      inject: [ConfigService],
    }),
```
I'm not exactly sure what's going on outside of the fact the types don't match.

**Comment by Alfagun74** [comment]
> I found a way to fix this issue although it may not be nice. 
> ```
> import { redisStore } from 'cache-manager-redis-store'
> 
> CacheModule.registerAsync({
>       isGlobal: true,
>       imports: [ConfigModule],
>       useFactory: async (configService: ConfigService) => ({
>         ttl: configService.get('CACHE_TTL'),
>         store: (await redisStore({
>           url: configService.get('REDIS_URL'),
>         })) as unknown as CacheStore,
>       }),
>       inject: [ConfigService],
>     }),
> ```
> I'm not exactly sure what's going on outside of the fact the types don't match.

Nah thanks imma stay on the old version until something proper gets released

**Comment by MikeRossXYZ** [comment]
@dougrocha You are effectively bypassing the Typescript type checks, which I would not recommend

I'm just stepping back into the Typescript and Node world again after a few years working in Go and Python, but I believe my #54  ⬆️ should get things working in NestJS by narrowing the type definitions provided by this package. 

If @dabroek approves of the changes, then you should be unblocked.

**Comment by iangregsondev** [comment]
Also there is another issue. Here is my solution. The in memory provider supports milliseconds by default but the redis provider (this one) supports seconds.

And as others have mentioned, the 3rd argument on the standard in-memory provider takes a number but this provider needs an object with key "ttl" and a number.

More info here..

```
    // It supports seconds and NOT milliseconds
    if (this.configService.cacheProvider == CacheProvider.redis) {
      // support seconds!!
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      await this.cacheManager.set(apiKey, key, { ttl: this.configService.cacheTtl / 1000 } as any);
    } else {
      // rest of providers should support milliseconds
      await this.cacheManager.set(apiKey, key, this.configService.cacheTtl);
    }

```

**Comment by iangregsondev** [comment]
My solution above is a bit messy, first I had to use "as any" to bypass type issue but should not all providers support milliseconds? Otherwise you need to add some logic as my reply above ^^^

**Comment by luluhoc** [comment]
@dabroek

**Comment by GabSnow24** [comment]
I managed to go through this error using require

When i used import and bypass using some weird thing in typescript i cant get anything in Redis container and was receiveng some errors in cache service. by using the code above, i managed to do things work.
`const redisStore = require('cache-manager-redis-store').redisStore;`

**Comment by cgat** [comment]
For NestJs 8, I wasn't able to get any of the solutions to work.

One solution shown in the NestJs __does__ work

```
CacheModule.register<ClientOpts>({
      store: redisStore,

      // Store-specific configuration:
      host: process.env.REDIS_HOST,
      port: parseInt(process.env.REDIS_PORT),
      password: process.env.REDIS_PASSWORD,
    }),
```

But you need to ensure that this is using `cache-manager-redis-store` 2.0.

**Comment by marianolinares** [comment]
The problem is redisStore is of type Store, and CacheStore is required. My solution is casting redisStore as CacheStore to fix it.

`CacheModule.register({
      store: redisStore as unknown as CacheStore,
      host: 'localhost',
      port: 6379,
    }),`

**Comment by Pangeran29** [comment]
Hello i solve this problem with down grade version of some packages, go check here
https://github.com/Pangeran29/testing-nestjs/blob/master/src/testing-cache-redis/testing-cahce-redis.MD

**Comment by yusuftaufiq** [comment]
Referring to [this comment](https://github.com/dabroek/node-cache-manager-redis-store/issues/40#issuecomment-1383193211), I think this issue can be solved if you use [node-cache-manager-redis-yet](https://github.com/node-cache-manager/node-cache-manager-redis-yet) instead.
```ts
import { CacheModule, Module } from '@nestjs/common';
import { redisStore } from 'cache-manager-redis-yet';

@Module({
  imports: [
    CacheModule.register({
      store: redisStore,
      url: 'redis://localhost:6379',
    }),
  ],
})

```

**Comment by puckj** [comment]
> Hello i solve this problem with down grade version of some packages, go check here https://github.com/Pangeran29/testing-nestjs/blob/master/src/testing-cache-redis/testing-cahce-redis.MD

Thank you king

**Comment by AhmedHdeawy** [comment]
> Hello i solve this problem with down grade version of some packages, go check here https://github.com/Pangeran29/testing-nestjs/blob/master/src/testing-cache-redis/testing-cahce-redis.MD

Thanks

**Comment by Exilz** [comment]
This package is riddled with issues, bad typing, missing methods that are not exposed and so on. I appreciate it for bridging nest's cache manager and redis but at this point I think it's easier to use something like `ioredis` with very basic abstractions for the set and get methods, and just be done with it.

This issue prevents me from properly closing the connection after running my end to end tests. Previously I couldn't set redis clusters properly because of the way the client is set up and exposed through this additional layer. It works great for simpler projects anyway.

**Comment by pratyush-prateek** [comment]
The solution mentioned by @cgat is working. If we use `cache-manager-redis-store` 2.0. It fixes the problem

**Comment by dev-whoan** [comment]
> Hello i solve this problem with down grade version of some packages, go check here
> https://github.com/Pangeran29/testing-nestjs/blob/master/src/testing-cache-redis/testing-cahce-redis.MD

Thanks It works.

**Comment by ram95krishh** [comment]
My solution with a dummy store fallback

```
import {
  CacheModule,
  CacheModuleOptions,
  CacheStore,
  Module
} from '@nestjs/common'
import { redisStore } from 'cache-manager-redis-store'

class DummyFallBackStore implements CacheStore {
  get(key: string) {
    console.debug(`Dummy Cache :: gets nothing for key - ${key}`)
    return Promise.resolve(undefined)
  }
  set(key: string) {
    console.debug(`Dummy Cache :: sets nothing for key - ${key}`)
    return Promise.resolve(undefined)
  }
  del(key: string) {
    logger.debug(`Dummy Cache :: deletes nothing for key - ${key}`)
    return Promise.resolve(undefined)
  }
}

@Module({
  imports: [
    CacheModule.registerAsync({
      useFactory: async () : Promise<CacheModuleOptions> => {
        try {
          const store = await redisStore({
            url: '<ConnectionString>'
          })
          return { store }
        } catch (error) {
          console.error(error)
          return { store: new DummyFallBackStore() }
        }
      }
    })
  ],
  controllers: [<MyController>]
})
export class MyModule {}
```

**Comment by CarsonRoscoe** [comment]
Can't take credit, but chiming in, this solution worked for me in app.module.

```ts
import { redisStore } from 'cache-manager-redis-store';
...
CacheModule.registerAsync < any > ({
    isGlobal: true,
    imports: [ConfigModule],
    useFactory: async (configService: ConfigService) => {
        const store = await redisStore({
            url: configService.get('REDIS_URL'),
            ttl: 60,
        });
        return {
            store: () => store,
        };
    },
    inject: [ConfigService],
}),
```

Do note, for some reason ttl changed from milliseconds to seconds with this change, despite the in-memory caching being in milliseconds with my environment

**Comment by nickytrung123** [comment]
This solution worked for me

```js
import { CacheModule, Module } from '@nestjs/common';
import type { RedisClientOptions } from 'redis';
// eslint-disable-next-line @typescript-eslint/no-var-requires
const redisStore = require('cache-manager-redis-store').redisStore;

@Module({
    imports: [
        CacheModule.registerAsync<RedisClientOptions>({
            useFactory: async () => ({
                store: await redisStore({
                    ttl,
                    host,
                    port,
                    password,
                }),
            }),
        }),
    ],
})

**Comment by MaksimKiselev** [comment]
## Problem

Example from official docs works only for `cache-manager-redis-store@^2.0.0`

`cache-manager-redis-store@^2.0.0` **incorrectly implement CacheManager interface** and accepts TTL as object where ttl property should be in seconds `{ ttl: TTL_IN_SECONDS }` see: https://github.com/dabroek/node-cache-manager-redis-store/issues/53#issuecomment-1325108666


## Solution

The [node-cache-manager](https://github.com/node-cache-manager/node-cache-manager#store-engines) have now official [redis](https://github.com/node-cache-manager/node-cache-manager-redis-yet)/[ioredis](https://github.com/node-cache-manager/node-cache-manager-ioredis-yet) cache stores.


`npm i --save cache-manager-redis-yet`

```
import { CacheModule, Module } from '@nestjs/common';
import { redisStore } from 'cache-manager-redis-yet';

@Module({
  imports: [
    CacheModule.registerAsync({
      isGlobal: true,
      useFactory: async () => ({
        store: await redisStore({
          host: process.env.REDIS_HOST,
          port: parseInt(process.env.REDIS_PORT!),
        }),
      }),
    }),
  ],
})
export class AppModule {}
```

**Comment by alexphamhp** [comment]
Thanks @MaksimKiselev 
I tried your solution and it works
But, I just want to note that
the TTL when I set, it is miliseconds not seconds

**Comment by MaksimKiselev** [comment]
@alexphamhp I meant that cache-manager-redis-store@^2.0.0 incorrectly implement CacheManager interface and accept ttl parameter as object with ttl property where ttl in seconds.

CacheManager interface define ttl parameter as number where ttl must be in microseconds.

I've update original comment. Thx. 🍻

**Comment by diariowebsquad** [comment]
this correct solution:
```
CacheModule.registerAsync<any>({
      isGlobal: true,
      imports: [ConfigModule],
      useFactory: async () => {
        const store = await redisStore({
          url: 'redis://redis_listings_service:6379',
          ttl: 0,
        });
        return {
          store: () => store,
        };
      },
    }),
```

**Comment by suoncha** [comment]
My solution for cache-manager v5 + cache-manager-redis-yet:
https://github.com/node-cache-manager/node-cache-manager/issues/210#issuecomment-1519127408

**Comment by tushermahmud** [comment]
Here is my solution. Node v:16.18.0 npm v:8.19.2

`"@nestjs/cache-manager": "^1.0.0",
    "@nestjs/common": "^9.0.0",
    "@nestjs/core": "^9.0.0",
    "@nestjs/jwt": "^10.0.3",
    "@nestjs/platform-express": "^9.0.0",
    "@nestjs/typeorm": "^9.0.1",
    "bcrypt": "^5.1.0",
    "bcryptjs": "^2.4.3",
    "cache-manager": "^5.2.1",
    "cache-manager-redis-store": "^2.0.0",
    "cache-manager-redis-yet": "^4.1.1",`



```
import {  Module } from '@nestjs/common';
import { CacheModule } from '@nestjs/cache-manager';
import { JwtModule } from '@nestjs/jwt';
import * as redisStore from 'cache-manager-redis-store';
@Module({
    imports:[JwtModule.register({ secret: 'secret', signOptions:{expiresIn:'1d'} }), CacheModule.register({
        store: redisStore as any,
        host: 'redis',
        port: 6379,
    })],
    exports:[JwtModule, CacheModule]
})
export class SharedModule {}
```

import { Body, Controller, Get, Post, Param, ParseIntPipe, Put, Delete, UseGuards, UseInterceptors, Inject } from '@nestjs/common';
import { AuthGuard } from 'src/auth/auth.guard';
import { ProductCreateDto } from './dtos/product-create.dto';
import { ProductService } from './product.service';
import { CacheKey,CacheInterceptor,CacheTTL, CACHE_MANAGER } from '@nestjs/cache-manager';
import {Cache} from 'cache-manager'

@Controller()
export class ProductController {
    constructor(private productService:ProductService, @Inject(CACHE_MANAGER) private cacheManager:Cache){
    }

    @UseGuards(AuthGuard)
    @Get('admin/products')
    async getAllProducts(){
        return await this.productService.findAll({});
    }

    @UseGuards(AuthGuard)
    @Post('admin/products')
    async create(@Body() body:ProductCreateDto){
        return await this.productService.save(body)
    }

    @UseGuards(AuthGuard)
    @Get('admin/products/:id')
    async getProduct(@Param('id', ParseIntPipe) id:number){
        return await this.productService.findById(id)
    }

    @UseGuards(AuthGuard)
    @Put('admin/products/:id')
    async update(@Param('id',ParseIntPipe) id:number, @Body() body:ProductCreateDto){
        await this.productService.update(id, body);
        return await this.productService.findById(id);
    }

    @UseGuards(AuthGuard)
    @Delete('admin/products/:id')
    async delete(@Param('id') id:number){
        await this.productService.delete(id);
    }

    @CacheKey('products_frontend')
    @CacheTTL(30*60)
    @UseInterceptors(CacheInterceptor)
    @Get('ambassador/products/frontend')
    async frontend(){
        return await this.productService.findAll({})
    }

    @Get('ambassador/products/backend')
    async backend(){
        let products = await this.cacheManager.get('products_backend');
        if(!products) {
            products = await this.productService.findAll({})
            await this.cacheManager.set('products_backend',products, 1800)
        }
        return products;
    }
}

**Comment by jongomes** [comment]
CacheModule.registerAsync({
      isGlobal: true,
      useFactory: async () => {

        try {
          const store = await redisStore({
            url: process.env.REDIS,
            ttl: 60000,
            fallback: 'memory' -- it would be interesting to fall back
          });
          return {store};
        } catch (error) {
          return {
            store: 'memory',
          };
        }
      },

**Comment by alhajee** [comment]
### cache.module.ts
I personally don't like having my configs directly in my App Module. I separate my configs into different modules and import them into `App module`.

You'd also notice that I prefer to read my environment variables from the `ConfigService` instead of reading directly from `process.env` (`This way you have more control for where env files are being read from`)

```typescript
import { Module } from '@nestjs/common';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { CacheModule, CacheModuleAsyncOptions } from '@nestjs/cache-manager';
import { redisStore } from 'cache-manager-redis-store';

@Module({
  imports: [
    CacheModule.registerAsync({
      isGlobal: true,
      imports: [ConfigModule], // No need to call ConfigModule.forRoot again
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => {
        const isActive = ['true', '1', 'yes'].includes(
          configService.get<string>('API_REDIS_STORE_IS_ACTIVE'),
        );
        return {
          store:
            isActive &&
            (await redisStore({
              // Store-specific configuration:
              socket: {
                host: configService.get<string>('API_REDIS_HOST'),
                port: +configService.get<number>('API_REDIS_PORT'),
              },
            })), // use redis when available or default to cache store
          ttl: 5000, // milliseconds
          max: 10, // maximum number of items in cache
        } as CacheModuleAsyncOptions;
      },
    }),
  ],
})
export class CacheConfigModule {}

```

**Comment by arlan85** [comment]
none of this works for me, I'm trying to get information from redis but I. don't get anything from the Redis database

**Comment by garryfreeman** [comment]
Guys, I had the same problem, I re-read the nestjs documentation and saw this note there

`cache-manager-redis-store does not support redis v4. In order for the ClientOpts interface to exist and work correctly you need to install the latest redis 3.x.x major release. See this issue to track the progress of this upgrade.`

```
npm i redis@3.1.2
```

```ts
CacheModule.register<RedisClientOptions>({
    store: redisStore,
    host: 'localhost',
    port: 6379,
}),
```

**Comment by LebranD** [comment]
here is my solution
 ```typescript
CacheModule.registerAsync<RedisClientOptions>({
            isGlobal: true,
            imports: [ConfigModule],
            inject: [ConfigService],
            useFactory: async (configService: ConfigService) => {
                const redis = configService.get('redis');
                return {
                    store: redisStore,
                    host: redis.host,
                    port: redis.port,
                } as StoreConfig;
            },
        }),
```

**Comment by cristian-moreno-ruiz** [comment]
It looks like this package is not maintained any more, and, although I'm facing the same issue at the moment, I moved to this other (which is a fork from this one but actively maintained):

https://github.com/node-cache-manager/node-cache-manager-redis-yet

Also, created this issue to be properly tracked:

https://github.com/node-cache-manager/node-cache-manager-redis-yet/issues/394

**Comment by janardhan5031** [comment]
I found a thing , when i save key value with TTL in redis from nestjs server after ttl expired, the key value not deleting.
Below is my code, let me know if anything went wrong.
Thank you.
`import { Module } from '@nestjs/common';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { CacheModule, CacheStore } from '@nestjs/cache-manager';
import { redisStore } from 'cache-manager-redis-store';

@Module({
  imports: [
    CacheModule.registerAsync({
      useFactory: async () => {
        const store = await redisStore({
          socket: {
            host: 'localhost',
            port: 6379,
          },
          password: '1234',
        });

        return {
          store: store as unknown as CacheStore,
        };
      },
    }),
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}

import { Controller, Get, Inject } from '@nestjs/common';
import { AppService } from './app.service';
import { Cache } from 'cache-manager';
import { CACHE_MANAGER } from '@nestjs/cache-manager';

@Controller()
export class AppController {
  constructor(
    private readonly appService: AppService,
    @Inject(CACHE_MANAGER) private cacheModule: Cache,
  ) {}

  @Get('health-check')
  async getHello() {
    const data = await this.cacheModule.get('HELLO');
    if (data) {
      return data;
    }

    await this.cacheModule.set('HELLO', 'Hello from store', 10);
    return 'Hello from server';
  }
}`

**Comment by NathBabs** [comment]
> none of this works for me, I'm trying to get information from redis but I. don't get anything from the Redis database

I am currently facing this issue, seems it is not using my redis but rather memory.

**Comment by phuphanvietlinkads** [comment]
I found solution:
```
  import * as redisStore from 'cache-manager-redis-yet';
 
  @Module({
    imports: [
      CacheModule.register({
        isGlobal: true,
        useFactory: () => ({
          store: redisStore,
        }),
        host: config.REDIS.HOST,
        port: config.REDIS.PORT,
      }),

**Comment by phankieuphu** [comment]
I found solution

```
  import * as redisStore from 'cache-manager-redis-yet';
 
  @Module({
    imports: [
      CacheModule.register({
        isGlobal: true,
        useFactory: () => ({
          store: redisStore,
        }),
        host: config.REDIS.HOST,
        port: config.REDIS.PORT,
      }),
```

**Comment by mariem2189** [comment]
I found a way to fix this issue, casting redisStore to 'any' to bypass the TypeScript error: 

```
import * as redisStore from 'cache-manager-redis-store';

@Module({
  imports: [
    CacheModule.register({
      store: redisStore as any,
      url: 'redis://localhost:6379',
    }),
  ],
})
```

**Comment by buidaoanhvan** [comment]
```javascript
import { Module } from '@nestjs/common';
import { CacheModule } from '@nestjs/cache-manager';
import { redisStore } from 'cache-manager-redis-yet';
import { AppController } from './app.controller';
import { AppService } from './app.service';

@Module({
  imports: [
    CacheModule.registerAsync({
      isGlobal: true,
      useFactory: async () => ({
        store: await redisStore({
          socket: {
            host: 'localhost',
            port: 6379
          }
        })
      })
    }),
  ],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule { }

```

**Comment by kgkg** [comment]
None of above solution works for me. It *seems* to work because in-memory cache is used silently instead of Redis. 

Try turning off Redis (image or service) and try to check if it even connects to it.

Working solution as of Nov 2024 with getting values from `ConfigModule`:

```ts
  imports: [
    CacheModule.registerAsync({
      isGlobal: true,
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService): Promise<CacheModuleOptions> => {
        try {
          const host = configService.get<string>('REDIS_HOST');
          const port = configService.get<number>('REDIS_PORT');
          const store = await redisStore({
            url: `redis://${host}:${port}`,
          });
          return { store };
        } catch (error) {
          console.error(error);
        }
      },
    }),
  ],
```

**Comment by zarej** [comment]
My working solution with [Keyv](https://keyv.org/docs/) and redis
```ts
import KeyvRedis from '@keyv/redis';
import { Module } from '@nestjs/common';
import { Cacheable, Keyv } from 'cacheable';

@Module({
  providers: [
    {
      provide: 'CACHE_INSTANCE',
      useFactory: () => {
        const secondary = new Keyv(new KeyvRedis({ url: 'redis://localhost:6379' }));
        if (!secondary) {
          throw new Error('Failed to create Keyv instance');
        }
        return new Cacheable({ secondary, ttl: '4h' });
      },
    },
  ],
  exports: ['CACHE_INSTANCE'],
})
export class CacheRedisModule {}
```
and in service you can use it with:
```ts
constructor(@Inject('CACHE_INSTANCE') private cache: Cacheable)
```

**Comment by marco-martins** [comment]
Like @kgkg mentioned, it seems to work because the in-memory cache is used silently instead of Redis. In my case, I only found out when I tried to persist the data with Redis and then realized that Redis was not in use either!

After a couple of hours, I got it working. In fact, we don’t need this package anymore, which is probably why it hasn’t been updated for a long time.

Following the [NestJS documentation](https://docs.nestjs.com/techniques/caching), the required packages are "@keyv/redis" and "cacheable".



Here is my implementation:

```javascript
import KeyvRedis from '@keyv/redis';
import {
  CacheModuleAsyncOptions,
  CacheModuleOptions,
} from '@nestjs/cache-manager';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { Cacheable } from 'cacheable';

export const CacheOptions: CacheModuleAsyncOptions = {
  isGlobal: true,
  imports: [ConfigModule],
  useFactory: async (
    configService: ConfigService,
  ): Promise<CacheModuleOptions> => {
    const host = configService.get<string>('REDIS_HOST');
    const port = configService.get<number>('REDIS_PORT');

    const redisStore = new Cacheable({
      primary: new KeyvRedis(`redis://${host}:${port}`),
      ttl: '10s',
    });

    return {
      stores: [redisStore],
    };
  },
  inject: [ConfigService],
};



```

```javascripts
@Module({
  imports: [
    CacheModule.registerAsync(CacheOptions),
 ...

```

Make sure that your imports are set correctly and are not imported from other libraries.

That was the only way it worked for me, even though the example in the NestJS docs didn’t work.

The only missing thing is that the code does not throw an exception if Redis is not running on your machine. Besides that, it works fine. 😉

**Comment by abhishekram404** [comment]
@marco-martins It worked flawlessly. Thanks.

**Comment by kgkg** [comment]
My current implementation with KeyV.
I'm using KeyV directly instead of Cacheable because I have multi-instance environment and I don't want any of them to store anything in their internal memory (I want Redis to be the only source of truth here). If it's not a case for you, you can just uncomment Cacheable line and comment the one above.

```ts
import { InternalServerErrorException, Module } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Keyv } from 'cacheable';
import KeyvRedis from '@keyv/redis';
import { RedisService } from './redis.service';

@Module({
  providers: [
    {
      provide: 'CACHE_INSTANCE',
      inject: [ConfigService],

      useFactory: async (configService: ConfigService) => {
        const host = configService.get<string>('REDIS_HOST');
        const port = configService.get<number>('REDIS_PORT');
        const tls = configService.get<string>('REDIS_USE_TLS') === 'true';

        const secondary = new KeyvRedis({
          socket: {
            host,
            port,
            tls,
            connectTimeout: 5000, // 5 seconds
          },
        });

        const ret = new Keyv(secondary);
        // const ret = new Cacheable({ secondary, ttl: '12h' });

        ret.on('error', (err) => {
          console.error('Connection Error', err);
          throw new InternalServerErrorException(
            `RedisService: Failed to connect to Redis server at ${host}:${port} with TLS=${tls}`,
          );
        });

        return ret;
      },
    },
    RedisService,
  ],
  exports: [RedisService],
})
export class RedisModule {}
```

**Comment by vvangwenkun** [comment]
This works for me in _**v3.0.1**_.

```javascript
import { Module } from '@nestjs/common';
import { CacheModule } from '@nestjs/cache-manager';
import { redisStore } from 'cache-manager-redis-store';
import { Keyv } from 'keyv';

@Module({
  imports: [
    CacheModule.registerAsync({
      useFactory: async () => {
        const store = await redisStore({
          host: 'localhost',
          port: 6379,
        });

        // It mainly addresses the method name compatibility issues between storage and Keyv
        Object.defineProperties(store, {
          delete: {
            value: store.del,
          },
          clear: {
            value: store.reset,
          },
        });

        return {
          stores: new Keyv({ store, namespace: '' }),
        };
      },
    }),
  ],
  providers: [],
  exports: [],
})
```
